### DISCUSSION TITLE
Modernize meteor tools

### DISCUSSION THREAD

**Discussion by afrokick** [meteor/meteor#12073] [Feature Requests]
Meteor as a build tool is a great software, but it should be modernise to simplify and reduce amount of code.

Because we use NodeJS 14.x under the hood, we can use things from ES2018+ like object rest/spread, [for-await-of loops](https://github.com/tc39/proposal-async-iteration), flat/flatMap etc.

Thanks to TypeScript, we can make the tools' codebase more safety with types and interfaces. (There were [PRs](https://github.com/meteor/meteor/pulls?q=is%3Apr+label%3AProject%3ATypeScript-conversion) already merged)

As a format tool, Prettier already configured.

So, the plan can be:

- apply format rules to all codebase in `tools`
- remove unused code (like https://github.com/meteor/meteor/pull/12072)
  - imports
  - constants
  - functions/classes
- remove `underscore`

- remove extensions from path in imports
- convert `require` to `import`

```diff
- var Progress = require('../console/progress.js').Progress;
+ import { Progress } from '../console/progress';
```

- provide types via TypeScript for general constants, utils etc.
- convert `exports.X = ...` to `export { X };`

@denihs @henriquealbert @fredmaiaarantes Any thoughts? Do you have a time to review such PRs?

**Comment by henriquealbert** [comment]
@afrokick It's a great initiative. We will do our best to review it and help everyone in the process.
The PRs should be as small as possible, please 🙏🏻  so it's easier to review and test each one.

Thank you 🙌🏻

**Comment by harryadel** [comment]
I've actually spoken to @zodern about this before and he said one it's one of the first things he'd like to do in order to lay the groundwork for more serious changes later on. 

>I think first we need to better document the Meteor Tool, refactor isobuild to make the high level process it follows clearer, and finish converting isobuild to typescript to save time for both me and anyone else who wants to contribute.

So please go ahead, you've already been doing great work to https://github.com/monti-apm/monti-apm-agent and with the blessings of Meteor team (@henriquealbert) it's a clear path for you :+1:
